### PR TITLE
find_sources: find build sources recursively

### LINKED ISSUE #8548: mypy doesn't recurse through directories, but -h says it does
According to `mypy -h`:
```
Mypy is a program that will type check your Python code.

Pass in any files or folders you want to type check. Mypy will
recursively traverse any provided folders to find .py files:

    $ mypy my_program.py my_src_folder
```

But it doesn't recurse, it only checks what's explicitly provided on the command line:
```
$ python3 -m venv venv
$ . venv/bin/activate
(venv) $ pip install mypy
...
(venv) $ mkdir -p test/test/
(venv) $ touch test/test/a.py
(venv) $ mypy test
There are no .py[i] files in directory 'test'
(venv) $ touch test/b.py
(venv) $ mypy test
Success: no issues found in 1 source file
(venv) $ mypy test test/test
Success: no issues found in 2 source files
```

The fix might be to change the documentation.  This is easy to workaround with `find`.

### PR DESCRIPTION
**Pull request by hauntsaninja**
Fixes #8548

This is important to fix because it's a really common source of
surprising false negatives.

I also lay some groundwork for supporting passing in namespace packages
on the command line. The approach towards namespace packages that this
anticipates is that if you pass in files to mypy and you want mypy to
understand your namespace packages, you'll need to specify explicit
package roots.

We also make many fewer calls to crawl functions, since we just pass
around what we need.

### CODE REVIEW

File: test-data/unit/cmdline.test [hunk 1]
```diff
@@ -68,6 +81,7 @@ undef
 undef
 [out]
 pkg/a.py:1: error: Name 'undef' is not defined
+pkg/subdir/a.py:1: error: Name 'undef' is not defined
```
Comment (JukkaL) [inline test-data/unit/cmdline.test:84]: Test importing from modules found via recursive directory traversal. Test cases where the directory contains `__init__.py` and where it doesn't exist.

Maybe test this also with `--namespace-packages`?

**Review by JukkaL — APPROVED** [review]
Thanks for fixing this bug! This was pretty bad. Looks good -- my only comments are about additional test cases.

Additional thanks for cleaning up some of the code, even if it wasn't needed for this PR.

File: test-data/unit/cmdline.test (hunk 1 above)
Comment (hauntsaninja) [reply test-data/unit/cmdline.test:84]: testCmdlinePackageSlash tests when the directory contains __init__.py and testCmdlinePackageContainingSubdir tests when it doesn't.
I added importing of the modules.

Comment (hauntsaninja) [reply test-data/unit/cmdline.test:88]: There currently isn't an interaction with `--namespace-packages` and cmdline tests are slow, so I won't do that unless told.
That said, I have another diff coming soon where I flesh out the package root code and make `--namespace-packages` the default. All these tests pass with that.

Comment (hauntsaninja) [reply test-data/unit/cmdline.test:88]: I'll also update the docs once all these changes (essentially important bug fixes + first three points of #8584) are shipped :-)

**Comment by intgr** [comment]
Ugh, turns out many users were relying on the old behavior and this causes issues because mypy now finds unexpected Python files from `node_modules`, virtualenv paths (`venv`), temporary directories etc. And there isn't a good way to exclude those paths (feature request #4675).

For example, creating an **empty virtualenv** with Python 3.9 **breaks mypy**:
```
% mkdir test
% cd test
% python3 -m venv ./venv
% echo 'a: str = 10' > test.py
% mypy .
venv/lib/python3.9/site-packages/setuptools/_distutils/command/bdist_msi.py:355: error: expected an indented block
Found 1 error in 1 file (errors prevented further checking)
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

I think this is a big enough regression that this change should be reverted in a hotfix release, until a longer-term solution can be figured out.

**Pull request merged**